src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+                                }
+                                replenishMain++;
+                                scalarEmission++;
+                                r--;
+                            }
+                            if (scalarEmission > 0) {
+                                r = producer.produced(scalarEmission);
+                            }
+                            if (r == 0L || o == null) {
+                                break;
+                            }
                         }
-                    } else {
-                        actual.onCompleted();
                     }
-                } else {
-                    actual.onCompleted();
-                }
-            }
-        }
 
-    }
-
-    private static final class MergeProducer<T> implements Producer {
-
-        private final MergeSubscriber<T> ms;
-
-        public MergeProducer(MergeSubscriber<T> ms) {
-            this.ms = ms;
-        }
-
-        private volatile long requested = 0;
-        @SuppressWarnings("rawtypes")
-        static final AtomicLongFieldUpdater<MergeProducer> REQUESTED = AtomicLongFieldUpdater.newUpdater(MergeProducer.class, "requested");
-
-        @Override
-        public void request(long n) {
-            if (requested == Long.MAX_VALUE) {
-                return;
-            }
-            if (n == Long.MAX_VALUE) {
-                requested = Long.MAX_VALUE;
-            } else {
-                BackpressureUtils.getAndAddRequest(REQUESTED, this, n);
-                if (ms.drainQueuesIfNeeded()) {
-                    boolean sendComplete = false;
-                    synchronized (ms) {
-                        if (ms.wip == 0 && ms.scalarValueQueue != null && ms.scalarValueQueue.isEmpty()) {
-                            sendComplete = true;
+                    /*
+                     * We need to read done before innerSubscribers because innerSubcribers are added
+                     * before done is set to true. If it were the other way around, we could read an empty
+                     * innerSubscribers, get paused and then read a done flag but an async producer
+                     * might have added more subscribers between the two.
+                     */
+                    boolean d = done;
+                    InnerSubscriber<?>[] inner = innerSubscribers;
+                    int n = inner.length;
+                    
+                    // check if upstream is done, there are no scalar values and no active inner subscriptions
+                    if (d && (svq == null || svq.isEmpty()) && n == 0) {
+                        Queue<Throwable> e = errors;
+                        if (e == null || e.isEmpty()) {
+                            child.onCompleted();
+                        } else {
+                            reportError();
+                        }
+                        if (svq != null) {
+                            svq.release();
```

Comment:
Yes, but who knows when that's going to happen due to the frequent use of `unsafeSubscribe` in the chains.